smt.go

```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
 	return proof, nil
 }
 
+// ProveClosest generates a SparseMerkleProof of inclusion for the first
+// key with the most common bits as the path provided.
+//
+// This method will follow the path provided until it hits a leaf node and then
+// exit. If the leaf is along the path it will produce an inclusion proof for
+// the key (and return the key-value internal pair) as they share a common
+// prefix. If however, during the tree traversal according to the path, a nil
+// node is encountered, the traversal backsteps and flips the path bit for that
+// depth (ie tries left if it tried right and vice versa). This guarentees that
+// a proof of inclusion is found that has the most common bits with the path
+// provided, biased to the longest common prefix
+func (smt *SMT) ProveClosest(path []byte) (
+	closestPath, closestValueHash []byte, // the closest leaf info for the key provided
+	proof *SparseMerkleProof, // proof of the key-value pair found
+	err error, // the error value encountered
+) {
+	workingPath := make([]byte, len(path))
+	copy(workingPath, path)
+	var siblings []treeNode
+	var sib treeNode
+	var parent treeNode
+	var depthDelta int
+
+	node := smt.tree
+	depth := 0
+	// continuously traverse the tree until we hit a leaf node
+	for depth < smt.depth() {
+		// save current node information as "parent" info
+		if node != nil {
+			parent = node
+		}
+		// resolve current node
+		node, err = smt.resolveLazy(node)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		// if we hit a nil node we backstep to the parent node and flip the path bit
+		// at the parent depth and select the other child
+		if node == nil {
+			node, err = smt.resolveLazy(parent)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			// trim the last sibling node added as it is no longer relevant
+			if len(siblings) > 0 {
+				siblings = siblings[:len(siblings)-1]
+			}
+			depth -= depthDelta
+			// flip the path bit at the parent depth
+			flipPathBit(workingPath, depth)
+		} else {
+			depthDelta = 0
+		}
+		// end traversal when we hit a leaf node
+		if _, ok := node.(*leafNode); ok {
+			break
+		}
+		if ext, ok := node.(*extensionNode); ok {
+			length, match := ext.match(workingPath, depth)
+			if match {
+				for i := 0; i < length; i++ {
+					siblings = append(siblings, nil)
+				}
+				depth += length
+				depthDelta += length
+				node = ext.child
+				node, err = smt.resolveLazy(node)
+				if err != nil {
+					return nil, nil, nil, err
+				}
+			} else {
+				node = ext.expand()
+			}
+		}
+		inner := node.(*innerNode)
```

Comment:
There is only 3 types of nodes if it is neither a leaf or an extension node it must be an inner node

Reply:
@Olshansk this actually poses a good question, this is only possible to fail when the tree is empty. We are looking for a proof of inclusion and the updated return here results in a successful proof of inclusion of an empty tree. This could be checked user side (ie, is closestPath a placeholder is so this tree was empty) but I think the new logic here should be good.

Reply:
Let's add a check. If it errors return an error that says "This can only happen if a tree is empty".

It'll be a good way to catch errors / edge cases we may not be aware of (which should THEORETICALLY not happen)